### PR TITLE
Batched grad for advanced indexing (index)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)
* **#47223 Batched grad for advanced indexing (index)**
* #47189 Fix sum batching rule, add simple clone batching rule
* #47188 Batching rules for complex view functions

This PR enables batched gradient computation for advanced indexing.
Previously, the backward formula was writing parts of the grad tensori
in-place to zeros_like(self). Since grad is a BatchedTensor and self is
not a BatchedTensor, this is not possible.

To solve the problem, we instead create a new tensor with
`grad.new_zeros` and then write to that in-place. This new tensor will
have the same batchedness as the `grad` tensor.

To prevent regressions (the autograd codegen special cases zeros_like
to avoid saving the `self` tensor for backward), we teach the autograd
codegen how to save `self.options()`.

Test Plan:
- new tests
- run old indexing tests

Differential Revision: [D24741684](https://our.internmc.facebook.com/intern/diff/D24741684)

### CODE REVIEW

**Review by albanD — APPROVED** [review]
lgtm!

Could you share a snippet of what the new codegen looks like just to make sure it works as intended?

**Comment by zou3519** [comment]
torch/csrc/autograd/generated/VariableTypeEverything.cpp:
```
Tensor index_Tensor(const Tensor & self, TensorList indices) {
  auto& self_ = unpack(self, "self", 0);
  auto indices_ = unpack(indices, "indices", 1);
  std::shared_ptr<IndexBackward> grad_fn;
  if (compute_requires_grad( self, indices )) {
    grad_fn = std::shared_ptr<IndexBackward>(new IndexBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self, indices ));
    grad_fn->self_sizes = self.sizes().vec();
    grad_fn->self_options = self.options();
    grad_fn->indices_ = make_saved_variable_list(indices);
    grad_fn->indices_size_ = indices.size();
  }
  ...
```
torch/csrc/autograd/generated/Functions.h:
```
struct TORCH_API IndexBackward : public TraceableFunction {
  using TraceableFunction::TraceableFunction;
  variable_list apply(variable_list&& grads) override;
  std::string name() const override { return "IndexBackward"; }
  void release_variables() override {
    std::lock_guard<std::mutex> lock(mutex_);
    indices_.clear();
    indices_released_ = true;
  }

  std::vector<int64_t> self_sizes;
  at::TensorOptions self_options;
  std::vector<SavedVariable> indices_;
  bool indices_released_ = false;
  size_t indices_size_;
};
```
torch/csrc/autograd/generated/Functions.cpp:
```
variable_list IndexBackward::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);
  TORCH_CHECK(!indices_released_, ERR_BACKWARD_TWICE);
  IndexRangeGenerator gen;
  auto self_ix = gen.range(1);
  auto indices_ix = gen.range(indices_size_);
  variable_list grad_inputs(gen.size());
  auto& grad = grads[0];
  auto indices = unpack_list(indices_);
  bool any_grad_defined = any_variable_defined(grads);
  if (should_compute_output({ indices_ix })) {
    auto grad_result = TensorList();
    copy_range(grad_inputs, indices_ix, grad_result);
  }
  if (should_compute_output({ self_ix })) {
    auto grad_result = any_grad_defined ? (index_backward(grad.new_zeros(self_sizes, self_options), indices, grad)) : Tensor();
    copy_range(grad_inputs, self_ix, grad_result);
  }
  return grad_inputs;
}
```

**Comment by jeffdaily** [comment]
@zou3519 and @albanD , I'd like help understanding this PR.  When I look at the generated code with and without this PR, it also affects `SplitBackward`, `UnsafeSplitBackward`, `SplitWithSizesBackward`, and `UnsafeSplitWithSizesBackward`.  Is this expected?

**Comment by zou3519** [comment]
@jeffdaily this PR causes the codegen saves "self.options()" for the backward pass. We didn't mean to change things outside of IndexBackward, but (I think) the change to the other Backward functions seems benign (could you post a code snippet of the before and after? I'd expect that in the "after", tensor.options() gets saved for backwards but doesn't need to).

Are you running into problems? It should be possible to special case the codegen to only save `self.options()` if `self` was not saved.

**Comment by jeffdaily** [comment]
We are seeing convergence issues in some models, and used git bisect to narrow it down to this PR.  I apologize for not having better details than that.  And fair warning, my knowledge of autograd, this codegen, etc, is still pretty limited.

I started with this upstream commit c9cebaf9b8f2ca90eaa5e974869a2f315b65354c, call this "before_revert", and then reverted this PR as "after_revert".  Here is a zip containing the generated autograd sources, organized as such.

[PR47223.zip](https://github.com/pytorch/pytorch/files/5954004/PR47223.zip)

The different files are as follows:

- Functions.cpp
- Functions.h
- VariableType_0.cpp
- VariableType_1.cpp
- VariableType_3.cpp
- VariableTypeEverything.cpp

One example difference is in `VariableTypeEverything.cpp`.  Does this PR change split such that each chunk is no longer a view of the original?  It causes `split_Tensor` to no longer use `SavedVariable`.

Current:

```
std::vector<Tensor> split_Tensor(const Tensor & self, int64_t split_size, int64_t dim) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  (void)_any_requires_grad;
  std::shared_ptr<SplitBackward> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<SplitBackward>(new SplitBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_sizes = self.sizes().vec();
    grad_fn->self_options = self.options();
    grad_fn->split_size = split_size;
    grad_fn->dim = dim;
  }
```

Reverted this PR
```
std::vector<Tensor> split_Tensor(const Tensor & self, int64_t split_size, int64_t dim) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  (void)_any_requires_grad;
  std::shared_ptr<SplitBackward> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<SplitBackward>(new SplitBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_sizes = self.sizes().vec();
    grad_fn->self_ = SavedVariable(self, false);
    grad_fn->split_size = split_size;
    grad_fn->dim = dim;
  }
```

**Comment by zou3519** [comment]
> Does this PR change split such that each chunk is no longer a view of the original?

No, this PR should not change view semantics. Each chunk is still a view of the original, we can check that via the following script:
```
import torch
x = torch.randn(100)
splits = x.split(5)
assert all(split._base is x for split in splits)
```
but also visually by modifing a split:
```
[ins] In [8]: import torch
         ...: x = torch.arange(10)
         ...: splits = x.split(5)
         ...: splits[0].zero_()
         ...: print(x)
tensor([0, 0, 0, 0, 0, 5, 6, 7, 8, 9])
```

> It causes split_Tensor to no longer use SavedVariable.

That seems fine to me. The [backward pass for split_Tensor](https://github.com/pytorch/pytorch/blob/1e171f024b13b797bcf4e8c8a055eb2dfd47850a/torch/csrc/autograd/FunctionsManual.cpp#L994-L1002) doesn't actually use the original `self` Tensor. By removing the SavedVariable here we may actually be saving some memory during training.

>  Here is a zip containing the generated autograd sources, organized as such.

Thank you for the zip, let me take a pass through for anything suspicious

**Comment by zou3519** [comment]
> Here is a zip containing the generated autograd sources, organized as such.

Looking through the zip everything looks fine / intentional. This PR makes it so that if a tensor isn't being used in backward (but the tensor's options are) then we no longer save the tensor, instead we save `tensor.options()`. If you could narrow down the convergence issue that would be helpful.

Feel free to continue asking questions!

**Comment by albanD** [comment]
Actually for these functions, it does have an important benefit: before we were saving `self` to be able to call `self.options()` during the backward. With this fix, we don't save `self` anymore. So that's one less Tensor that is kept around for the backward :)

**Comment by jeffdaily** [comment]
> Actually for these functions, it does have an important benefit: before we were saving `self` to be able to call `self.options()` during the backward. With this fix, we don't save `self` anymore. So that's one less Tensor that is kept around for the backward :)

Does that imply that this PR may have revealed a Tensor lifetime bug in ROCm?  One less Tensor ref might mean a Tensor is reclaimed too early by the caching allocator?

**Comment by zou3519** [comment]
> Does that imply that this PR may have revealed a Tensor lifetime bug in ROCm? One less Tensor ref might mean a Tensor is reclaimed too early by the caching allocator?

Knowing nothing about ROCm, I feel like there would be bigger problems (e.g. crashes) if there are lifetime bugs.
